### ISSUE TITLE
[BUG] (unknown cause) Volume thinks it is expanding (e.g. 20gb to 10gb)

### ISSUE THREAD

**Issue by voarsh2** [longhorn/longhorn#4181]
## Describe the bug
I updated my RKE cluster config and many of my volumes were being rebuilt.
Some volumes were corrupted and needed restoring.

One volume (I've seen this a few times) thinks it is in expanding state:

![image](https://user-images.githubusercontent.com/60437695/175850886-39f9ef61-061a-4ec1-9fe0-6372636c8433.png)

## Environment

 - Longhorn version: 1.2.4
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Rancher Catalog App
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 3
 - Node config
   - OS type and version: Ubuntu 21
   - CPU per node: 10 variable
   - Memory per node: 20GB variable
   - Disk type(e.g. SSD/NVMe): HDD
   - Network bandwidth between the nodes: 1Gbe
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): KVM
 - Number of Longhorn volumes in the cluster: 50~

**Comment by voarsh2** [comment]
This has happened to two volumes when I opened this issue. The size gets updated in LH somehow... but the PVC/PV show "10 GB) in size... and when I delete and restore the volume the new size is 20GB.

**Comment by mantissahz** [comment]
@voarsh2 we want to know how did you update the RKE cluster and what configurations you modified, Did you do anything before and when updating the cluster configuration?
If yes, could you provide the detail steps about what you did?
Thanks.

**Comment by gcarrarom** [comment]
I'm running into this exact issue right now. I have a few volumes that started resizing on their own and now they're stuck on "expanding" even though I haven't requested it. It scares me as it shows only one replica for those:
![CleanShot 2023-07-24 at 11 29 54](https://github.com/longhorn/longhorn/assets/10549675/cad11fbd-3f06-4f28-92f2-9f94656e9984)
Trying to cancel the expansion fails with the following error:

> unable to cancel expansion for volume pvc-067cd910-1e64-4a33-af0c-c9f404e33704: volume expansion is not started

I luckily have a maintenance window tonight that I could use to restore from backup. But I would like to understand the cause of this.

**Comment by derekbit** [comment]
What's your LH version? 
Probably related to https://github.com/longhorn/longhorn/issues/5513

cc @ejweber

**Comment by gcarrarom** [comment]
> What's your LH version? Probably related to #5513
> 
> cc @ejweber

Sorry, I was editing to add the info. I'll check this issue and report back.

**Comment by gcarrarom** [comment]
Sure, but problem is now solved by doing the aforementioned workaround. But I will share the support bundle as soon as it happens again.

Let me know if you'd like the support bundle regardless. Although the pods are all restarted at this point - I'm not sure what is packaged onto this bundle.

**Comment by ejweber** [comment]
Hello all.

Yes, unfortunately this still has the chance of occurring in v1.5.0. There is an additional fix in v1.4.3 and in v1.5.1 that eliminates its reproducibility on my side, but it may be possible to reproduce in a different way than I have done so far.

See the tickets below for much more context, but for a quick recap:

- We have only ever managed to reproduce the issue when an instance-manager pod goes down with processes still running inside of it. In environments where this happens frequently (e.g. the user shuts down cluster nodes every night without scaling down workloads first), it can be relatively common.
- The root cause is a replica being expanded by the wrong engine during the churn of an instance-manager or node shutdown/restart. Replica addresses change quickly during this churn, and until recently we had not developed a solid way to prevent the small chance of a miscommunication.
- The replica is only "too large" on disk. The longhorn-system still considers it to have its original size. A replica can even run this way indefinitely without issue because the engine simply writes to offsets that should be available to it.
- Eventually, other replicas are shut down and require rebuilding. This is completely normal in Longhorn when a node is taken offline, etc. However, a new replica cannot be correctly rebuilt from the replica that is too large.
- When circumstances are right, we can get into a situation in which the only remaining healthy replica has the wrong size. This is usually the moment a user might recognize something bad has happened or is happening.

Additional context:

- Issue tracking the fix for v1.4.2 and v1.5.0 (which was not sufficient); https://github.com/longhorn/longhorn/issues/5709.
- Issue tracking the continued occurrence in v1.4.2 (similar to v1.5.0): https://github.com/longhorn/longhorn/issues/6078.
- Issue tracking the fix for v1.4.3 and v1.5.1: https://github.com/longhorn/longhorn/issues/6217.
- Issue tracking the long-term fix that should make this impossible: https://github.com/longhorn/longhorn/issues/5845.

**Comment by ejweber** [comment]
> EDIT2: Just did a trial run on a couple of workloads that are not production, it works now. Here's what I did:
> 
> 1. Backup the volume
> 2. Delete the volume
> 3. Recover from Backup
> 4. Let the workload restart

@gcarrarom, this is interesting to me. Normally we recommend simply expanding the volume to the larger size as the workaround (e.g. intentionally expand the volume to 10 GiB from 8 GiB).

I think backing up the volume in this state and then restoring from backup would implicitly cause this volume expansion. Did you end up with a working volume that was larger than expected or the original size?

**Comment by derekbit** [comment]
I think we need a KB to explain the unexpected expansion and the workaround. WDYT? @ejweber @innobead

**Comment by gcarrarom** [comment]
> > EDIT2: Just did a trial run on a couple of workloads that are not production, it works now. Here's what I did:
> > 
> > 1. Backup the volume
> > 2. Delete the volume
> > 3. Recover from Backup
> > 4. Let the workload restart
> 
> @gcarrarom, this is interesting to me. Normally we recommend simply expanding the volume to the larger size as the workaround (e.g. intentionally expand the volume to 10 GiB from 8 GiB).
> 
> I think backing up the volume in this state and then restoring from backup would implicitly cause this volume expansion. Did you end up with a working volume that was larger than expected or the original size?

Not really, I just recovered it with the original size. Nothing changed.

**Comment by Eilyre** [comment]
@gcarrarom Sadly I have this issue after upgrading to v1.5.1 from v1.4.2. I have random PV's in this state after upgrade (27 volumes out of 130).
